vdr/store/bbolt.go

```diff
@@ -0,0 +1,305 @@
+				return err
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (store *bboltStore) filterDocument(doc *documentVersion, metadata *vdr.ResolveMetadata) error {
+	// Verify deactivated
+	if IsDeactivated(doc.Document) {
+		return vdr.ErrDeactivated
+	}
+
+	if metadata == nil {
+		return nil
+	}
+
+	// Verify creation and update time
+	if metadata.ResolveTime != nil {
+		resolveTime := *metadata.ResolveTime
+
+		if doc.Metadata.Created.After(resolveTime) {
+			return vdr.ErrNotFound
+		}
+
+		if doc.Metadata.Updated != nil {
+			if doc.Metadata.Updated.After(resolveTime) {
+				return vdr.ErrNotFound
+			}
+		}
+	}
+
+	// Verify KeyTransaction
+	if metadata.SourceTransaction != nil {
+		for i, keyTx := range doc.Metadata.SourceTransactions {
+			if keyTx.Equals(*metadata.SourceTransaction) {
+				break
+			}
+
+			if i == len(doc.Metadata.SourceTransactions)-1 {
+				return vdr.ErrNotFound
+			}
+		}
+	}
+
+	return nil
+}
+
+// Resolve returns the DID Document for the provided DID
+func (store *bboltStore) Resolve(id did.DID, metadata *vdr.ResolveMetadata) (document *did.Document, documentMeta *vdr.DocumentMetadata, txErr error) {
+	txErr = store.db.View(func(tx *bbolt.Tx) error {
+		documents := tx.Bucket(documentsBucket)
+		if documents == nil {
+			return vdr.ErrNotFound
+		}
+
+		// If `metadata.Hash` is set we can fetch the document without iterating
+		if metadata != nil && metadata.Hash != nil {
+			doc, err := store.getDocumentVersion(documents, id, *metadata.Hash)
+			if err != nil {
+				return nil
+			}
+
+			if err := store.filterDocument(doc, metadata); err != nil {
+				return err
+			}
+
+			document = &doc.Document
+			documentMeta = &doc.Metadata
+
+			return nil
+		}
+
+		versions := tx.Bucket(versionsBucket)
+		if versions == nil {
+			return vdr.ErrNotFound
+		}
+
+		versionList := &documentVersionList{}
+
+		data := versions.Get([]byte(id.String()))
+		if data == nil {
+			return vdr.ErrNotFound
+		}
+
+		if err := json.Unmarshal(data, versionList); err != nil {
+			return err
+		}
+
+		for _, versionHash := range versionList.Versions {
+			data = documents.Get(store.getDocumentKey(id, versionHash))
+			if data == nil {
+				continue
+			}
+
+			doc := &documentVersion{}
+
+			if err := json.Unmarshal(data, doc); err != nil {
+				return err
+			}
+
+			if err := store.filterDocument(doc, metadata); err == nil {
+				document = &doc.Document
+				documentMeta = &doc.Metadata
+
+				return nil
+			}
+		}
+
+		return vdr.ErrNotFound
+	})
+
+	return
+}
+
+// Write writes a DID Document
+func (store *bboltStore) Write(document did.Document, metadata vdr.DocumentMetadata) error {
+	return store.db.Update(func(tx *bbolt.Tx) error {
+		versions, err := tx.CreateBucketIfNotExists(versionsBucket)
+		if err != nil {
+			return err
+		}
+
+		key := []byte(document.ID.String())
+
+		if versions.Get(key) != nil {
+			return vdr.ErrDIDAlreadyExists
```

Comment:
This will sting when rewalking the DAG

Reply:
We can also return `nil` here instead? In the sense that nothing really went wrong.